### ISSUE TITLE
Support gradient clipping before & after allreduce

### ISSUE THREAD

**Issue by alsrgv** [horovod/horovod#1346]
**Is your feature request related to a problem? Please describe.**
Currently, each framework allows one (not always consistent) way to do gradient clipping. Generally, one could do gradient clipping locally (before allreduce) or globally (after allreduce).

In PyTorch, gradient clipping is done globally (after allreduce).

In MXNet and TensorFlow (since Horovod 0.17.0), gradient clipping is done locally (before allreduce).

**Describe the solution you'd like**
We should allow the user to specify gradient clipping and other gradient transformations both before and after allreduce.  This can easily be done with either a couple of Lambdas passed to `hvd.DistributedOptimizer()`, or by implementing a callback mechanism with the list of callbacks similarly passed to `hvd.DistributedOptimizer()`.

The latter approach has the benefit of being more generalizable and extensible in the future.

cc @DEKHTIARJonathan
